core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.uint8_tVar
+import kotlin.native.ref.WeakReference
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink
+) : NSOutputStream(toMemory = Unit), NSStreamDelegateProtocol {
```

Comment:
I'm not sure how things work under the hood when inheriting NSInputStream/NSOutputStream and calling a non-default initializer (I understand that no-arg initializer is not available here) as according to Apple's docs, these initializers aimed to instantiate some specific stream's subclass.

Reply:
This was the only way I found to appease the requirement `Unable to call non-designated initializer as super constructor`, which is the error the compiler throws if you use the no-arg constructor. This seems to be a conflict between the way Kotlin and Objective-C object inheritance works, where Objective-C can return a subclass implementation as `self` from an `init` function.

I'm not sure there really should be an expected designated initializer in Kotlin for either of these classes. This is likely an interop error. In Objective-C this compiles and I'm able to construct `[[MyNSInputStream alloc] init]` just fine:

```
// MyNSInputStream.h
@interface MyNSInputStream : NSInputStream
- (instancetype)init;
@end

// MyNSInputStream.m
@implementation MyNSInputStream
- (instancetype)init {
    self = [super init];
    return self;
}
@end
```

It works to use either the memory or URL constructor of both `NSInputStream` and `NSOutputStream`. I figured the empty memory versions were likely to be the most lightweight. Is there some way to suppress this designated initializer requirement to call the no-arg constructor? It might also make sense to file an issue to fix the interop. Since we're overriding all of `NSInputStream` and `NSOutputStream`'s public API functions and not using anything from the superclass, I think this should be safe though.

Reply:
Related issues:

https://youtrack.jetbrains.com/issue/KT-47992
https://github.com/JetBrains/kotlin-native/issues/2010

Looks like there is a workaround for disabling in custom .def interop, but this won't work for the platform interop in this case.

Reply:
In fact, I was mostly concerned about the possibility of inheriting some methods from NSInput/OutputStream subclasses instantiated for `initFromData`, `initToMemory`. But it's definitely not the case, as there were errors due to non-implemented run-loop-related methods previously.

Reply:
I added the suggested doc comments. It would be strange for a consumer to call any of those `init` functions after receiving the instantiated stream object from `asNSInputStream()` or `asNSOutputStream()`. In Kotlin it already shows an error trying to call `init` functions directly, not through a constructor. I don't think I've ever seen a call to an `init` function even in Objective-C that wasn't directly after a call to `alloc`. Swift doesn't allow this either, where `init` functions are only usable to construct objects now.